raiden_contracts/data/source/raiden/TokenNetwork.sol

```diff
@@ -475,7 +502,74 @@ contract TokenNetwork is Utils {
         // balance proof in storage. This should never fail as we use isOpen.
         assert(participant_state.nonce == 0);
         assert(partner_state.nonce == 0);
+    }
+
+    /// @notice Cooperatively settles the balances between the two channel
+    /// participants and transfers the agreed upon token amounts to the
+    /// participants. After this the channel lifecycle has ended and no more
+    /// operations can be done on it.
+    /// An important constraint is that this function checks that all tokens
+    /// in this channel are withdrawn. This means that the channel can *not*
+    /// have any outstanding locked transfers.
+    /// @param channel_identifier Identifier for the channel on which this
+    /// operation takes place
+    /// @param data1 Withdraw data of the participant initiating the cooperative settlement
+    /// @param data2 Withdraw data of the second participant
+    function cooperativeSettle(
+        uint256 channel_identifier,
+        WithdrawInput memory data1,
+        WithdrawInput memory data2
+    )
+        external
+        isOpen(channel_identifier)
+    {
+        uint256 total_deposit;
+        bytes32 pair_hash;
+
+        // Validate that authenticated partners and the channel identifier match
+        require(channel_identifier == getChannelIdentifier(data1.participant, data2.participant));
+        Channel storage channel = channels[channel_identifier];
+
+        Participant storage participant1_state = channel.participants[data1.participant];
+        Participant storage participant2_state = channel.participants[data2.participant];
+        total_deposit = participant1_state.deposit + participant2_state.deposit;
 
+        // The sum of the provided balances must be equal to the total
+        // available deposit. This also implies that no locks must exist in the channel
+        // when this is called, as otherwise the withdrawable amount would be smaller
+        // than required.
+        require((data1.total_withdraw + data2.total_withdraw) == total_deposit, "CS: incomplete amounts");
+        // Overflow check for the balances addition from the above check.
+        // This overflow should never happen if the token.transfer function is implemented
+        // correctly. We do not control the token implementation, therefore we add this
+        // check for safety.
+        require(data1.total_withdraw <= data1.total_withdraw + data2.total_withdraw);
+
+        if (data1.total_withdraw > 0) {
+            this.setTotalWithdraw(
+                channel_identifier,
+                data1.participant,
+                data1.total_withdraw,
+                data1.expiration_block,
+                data1.participant_signature,
+                data1.partner_signature
```

Comment:
Nitpick question: can't we call `setTotalWithdraw2` directly with `data1` here? And data2 below? Wouldn't this be cheaper in gas?

Reply:
Well spotted!